## From Latex

In [36]:
from TexSoup import TexSoup
import regex as re
import os
import nbformat as nbf

In [10]:
with open("pml1.lof") as fp:
    LoF_File_Contents = fp.read()

In [11]:
LoF_File_Contents
soup = TexSoup(LoF_File_Contents)

In [12]:
def extract_script_from_latex(caption):
    py_pattern = r"\{\S+?\.py\}"
    ipynb_pattern = r"\{\S+?\.ipynb\}"

    matches = re.findall(py_pattern, str(caption)) + re.findall(ipynb_pattern, str(caption))
    extracted_scripts = []
    for each in matches:
        if "https" not in each:
            each = each.replace("{", "").replace("}", "").replace("\\_", "_")
            extracted_scripts.append(each)

    return extracted_scripts

In [13]:
# create mapping of fig_no to list of script_name
py_pattern = r"\{\S+?\.py\}"
ipynb_pattern = r"\{\S+?\.ipynb\}"
fig_cnt = 0
script_name_dict = {}
for fig_line in soup.find_all("numberline"):
    fig_no = str(fig_line.contents[0])
    extracted_scripts = extract_script_from_latex(str(fig_line))
    if extracted_scripts != []:
        script_name_dict[fig_no] = extracted_scripts

script_name_dict

{'1.3': ['iris_plot.py'],
 '1.4': ['iris_dtree.ipynb'],
 '1.5': ['linreg_residuals_plot.py'],
 '1.6': ['linreg_2d_surface_demo.py'],
 '1.7': ['linreg_poly_vs_degree.py'],
 '1.8': ['iris_kmeans.py'],
 '1.9': ['iris_pca.py'],
 '1.12': ['mnist_viz_tf.py', 'emnist_viz_jax.ipynb'],
 '1.13': ['fashion_viz_tf.py', 'cifar_viz_tf.py'],
 '2.1': ['discrete_prob_dist_plot.ipynb'],
 '2.2': ['gauss_plot.py', 'quantile_plot.py'],
 '2.4': ['bimodal_dist_plot.py'],
 '2.5': ['anscombes_quartet.ipynb'],
 '2.6': ['datasaurus_dozen.ipynb'],
 '2.9': ['binom_dist_plot.ipynb'],
 '2.10': ['activation_fun_plot.py'],
 '2.11': ['iris_logreg.py'],
 '2.12': ['softmax_plot.py'],
 '2.13': ['iris_logreg.py'],
 '2.14': ['linreg_1d_hetero_tfp.py'],
 '2.15': ['student_laplace_pdf_plot.py'],
 '2.16': ['robust_pdf_plot.py'],
 '2.17': ['beta_dist_plot.py', 'gamma_dist_plot.py'],
 '2.23': ['centralLimitDemo.py'],
 '2.24': ['change_of_vars_demo1d.py'],
 '3.3': ['simpsons_paradox.ipynb'],
 '3.5': ['gauss_plot_2d.py'],
 '3.6': 

In [14]:
# create mapping between fig_no and actual_url path
def make_url_from_fig_no_and_script_name(
    fig_no, script_name, base_url="https://github.com/probml/pyprobml/blob/master/notebooks", book_no=1
):
    chapter_no = int(fig_no.strip().split(".")[0])
    base_url_ipynb = os.path.join(base_url, f"book{book_no}/{chapter_no:02d}")
    if ".py" in script_name:
        script_name = script_name[:-3] + ".ipynb"
    return os.path.join(base_url_ipynb, script_name)

In [17]:
base_url_test = "https://github.com/karm-patel/pyprobml-1/tree/dead_urls/notebooks"
base_url = "https://github.com/probml/pyprobml/blob/master/notebooks"

fig_no_urls_mapping = {}
for fig_no in script_name_dict:
    script_names = script_name_dict[fig_no]
    fig_no_urls_mapping[fig_no] = [
        make_url_from_fig_no_and_script_name(fig_no, script_name, base_url=base_url) for script_name in script_names
    ]
fig_no_urls_mapping

{'1.3': ['https://github.com/probml/pyprobml/blob/master/notebooks/book1/01/iris_plot.ipynb'],
 '1.4': ['https://github.com/probml/pyprobml/blob/master/notebooks/book1/01/iris_dtree.ipynb'],
 '1.5': ['https://github.com/probml/pyprobml/blob/master/notebooks/book1/01/linreg_residuals_plot.ipynb'],
 '1.6': ['https://github.com/probml/pyprobml/blob/master/notebooks/book1/01/linreg_2d_surface_demo.ipynb'],
 '1.7': ['https://github.com/probml/pyprobml/blob/master/notebooks/book1/01/linreg_poly_vs_degree.ipynb'],
 '1.8': ['https://github.com/probml/pyprobml/blob/master/notebooks/book1/01/iris_kmeans.ipynb'],
 '1.9': ['https://github.com/probml/pyprobml/blob/master/notebooks/book1/01/iris_pca.ipynb'],
 '1.12': ['https://github.com/probml/pyprobml/blob/master/notebooks/book1/01/mnist_viz_tf.ipynb',
  'https://github.com/probml/pyprobml/blob/master/notebooks/book1/01/emnist_viz_jax.ipynb'],
 '1.13': ['https://github.com/probml/pyprobml/blob/master/notebooks/book1/01/fashion_viz_tf.ipynb',
  'ht

### Multinotebooks

In [18]:
more_than_one = 0
multi_notebooks = {}
for each in fig_no_urls_mapping:

    if len(fig_no_urls_mapping[each]) > 1:
        multi_notebooks[each] = []
        for url in fig_no_urls_mapping[each]:
            multi_notebooks[each].append(url)
        more_than_one += 1
#         print()
more_than_one
multi_notebooks

{'1.12': ['https://github.com/probml/pyprobml/blob/master/notebooks/book1/01/mnist_viz_tf.ipynb',
  'https://github.com/probml/pyprobml/blob/master/notebooks/book1/01/emnist_viz_jax.ipynb'],
 '1.13': ['https://github.com/probml/pyprobml/blob/master/notebooks/book1/01/fashion_viz_tf.ipynb',
  'https://github.com/probml/pyprobml/blob/master/notebooks/book1/01/cifar_viz_tf.ipynb'],
 '2.2': ['https://github.com/probml/pyprobml/blob/master/notebooks/book1/02/gauss_plot.ipynb',
  'https://github.com/probml/pyprobml/blob/master/notebooks/book1/02/quantile_plot.ipynb'],
 '2.17': ['https://github.com/probml/pyprobml/blob/master/notebooks/book1/02/beta_dist_plot.ipynb',
  'https://github.com/probml/pyprobml/blob/master/notebooks/book1/02/gamma_dist_plot.ipynb'],
 '4.14': ['https://github.com/probml/pyprobml/blob/master/notebooks/book1/04/dirichlet_3d_triangle_plot.ipynb',
  'https://github.com/probml/pyprobml/blob/master/notebooks/book1/04/dirichlet_3d_spiky_plot.ipynb'],
 '4.20': ['https://gith

### Caption reading

In [20]:
def convert_to_ipynb(file):
    if ".py" in file:
        return file[:-3] + ".ipynb"
    return file

In [34]:
# create mapping of fig_no to list of script_name
py_pattern = r"\{\S+?\.py\}"
ipynb_pattern = r"\{\S+?\.ipynb\}"

whole_link_ipynb = r"\{\S+\.ipynb\}"
whole_link_py = r"\{\S+\.py\}"

fig_cnt = 0
cleaned_caption = {}
for fig_line in soup.find_all("numberline"):
    fig_no = str(fig_line.contents[0])

    # if it does not contain multi_notebooks
    if fig_no not in multi_notebooks:
        continue

    fig_line = (
        str(fig_line)
        .replace(r"\ignorespaces", "")
        .replace(r" \relax", "")
        .replace(r"\href", "")
        .replace(r"\url", "")
        .replace(r'\cc@accent {"705E}', "")
        .replace(r"\numberline", "")
    )

    # print(fig_no, end=" ")
    links = re.findall(whole_link_ipynb, str(fig_line)) + re.findall(whole_link_py, str(fig_line))
    # print(fig_no, links)
    for link in links:
        script = extract_script_from_latex(link)[0]
        script_ipynb = convert_to_ipynb(script)
        original_url = f"[{script_ipynb}]({make_url_from_fig_no_and_script_name(fig_no,script_ipynb)})"
        fig_line = fig_line.replace(link, original_url)

    fig_line = fig_line.split("}{")[1][:-2]
    # print(fig_no, fig_line, end="\n\n")
    cleaned_caption[fig_no] = fig_line


cleaned_caption

{'1.12': ' (a) Visualization of the MNIST dataset. Each image is $28 \\times 28$. There are 60k training examples and 10k test examples. We show the first 25 images from the training set. Generated by [mnist_viz_tf.ipynb](https://github.com/probml/pyprobml/blob/master/notebooks/book1/01/mnist_viz_tf.ipynb). (b) Visualization of the EMNIST dataset. There are 697,932 training examples, and 116,323 test examples, each of size $28 \\times 28$. There are 62 classes (a-z, A-Z, 0-9). We show the first 25 images from the training set. Generated by [emnist_viz_jax.ipynb](https://github.com/probml/pyprobml/blob/master/notebooks/book1/01/emnist_viz_jax.ipynb).',
 '1.13': ' (a) Visualization of the Fashion-MNIST dataset \\citep{fashion}. The dataset has the same size as MNIST, but is harder to classify. There are 10 classes: T-shirt/top, Trouser, Pullover, Dress, Coat, Sandal, Shirt, Sneaker, Bag, Ankle-boot. We show the first 25 images from the training set. Generated by [fashion_viz_tf.ipynb](ht

### Write caption to dummy ipynb file

In [37]:
# create new notebook and add caption to it
# https://stackoverflow.com/questions/38193878/how-to-create-modify-a-jupyter-notebook-from-code-python
book_no = 1
for fig_no in cleaned_caption:

    # make relative path for new dummy notebook
    chapter_no = int(fig_no.split(".")[0])
    relative_path = "../notebooks/"
    dummpy_notebook = f"fig_{fig_no.replace('.','_')}.ipynb"
    fig_path = os.path.join(relative_path, f"book{book_no}/{chapter_no:02d}", dummpy_notebook)
    print(fig_path.split("/")[-1], end="\n")

    nb = nbf.v4.new_notebook()
    nb["cells"] = [nbf.v4.new_markdown_cell(cleaned_caption[fig_no])]
    with open(fig_path, "w") as f:
        nbf.write(nb, f)

fig_1_12.ipynb
fig_1_13.ipynb
fig_2_2.ipynb
fig_2_17.ipynb
fig_4_14.ipynb
fig_4_20.ipynb
fig_5_2.ipynb
fig_5_10.ipynb
fig_8_1.ipynb
fig_8_14.ipynb
fig_8_26.ipynb
fig_11_10.ipynb
fig_11_19.ipynb
fig_18_4.ipynb
fig_20_24.ipynb
fig_20_25.ipynb
fig_20_26.ipynb
fig_20_30.ipynb
fig_20_31.ipynb
fig_20_33.ipynb
fig_20_36.ipynb
fig_20_37.ipynb
fig_20_38.ipynb
fig_20_41.ipynb
fig_21_11.ipynb


## From pdf 

In [71]:
import os
import regex as re

try:
    import fitz  # this is pymupdf
except:
    %pip install fitz
import requests

In [3]:
with fitz.open("pml1-listoffigures.pdf") as doc:
    text = ""
    for page in doc:
        text += page.get_text()

# print(text)

In [4]:
pattern = r"[\n][0-9]+[.][0-9]+[\n]"
fig_numbers = re.findall(pattern, text)
len(fig_numbers)

431

In [5]:
captions = re.compile(pattern).split(text)[1:]
len(captions)

431

In [144]:
a = "acfd.py"
a[:-3]

'acfd'

In [97]:
seperator_pattern = r"[\s\n]"
scripts = {}
book_no = 1
for no, cap in zip(fig_numbers, captions):
    words = re.compile(seperator_pattern).split(cap)
    fig_no = no.strip()
    chapter_no = int(fig_no.split(".")[0])
    base_url_ipynb = os.path.join(base_url_test, f"book{book_no}/{chapter_no:02d}")
    script_names = []

    for each in words:
        if ".py" in each:
            script_name = each[: each.find(".py")]
            script_names.append(os.path.join(base_url_ipynb, script_name + ".ipynb"))
        elif ".ipynb" in each:
            script_name = each[: each.find(".ipynb")]
            script_names.append(os.path.join(base_url_ipynb, script_name + ".ipynb"))
    print(fig_no, script_names)
    scripts[fig_no] = list(set(script_names))  # remove duplicate scripts in same figure

1.1 []
1.2 []
1.3 ['https://github.com/karm-patel/pyprobml-1/tree/dead_urls/notebooks/book1/01/iris_plot.ipynb']
1.4 ['https://github.com/karm-patel/pyprobml-1/tree/dead_urls/notebooks/book1/01/iris_dtree.ipynb']
1.5 ['https://github.com/karm-patel/pyprobml-1/tree/dead_urls/notebooks/book1/01/linreg_residuals_plot.ipynb']
1.6 ['https://github.com/karm-patel/pyprobml-1/tree/dead_urls/notebooks/book1/01/linreg_2d_surface_demo.ipynb']
1.7 ['https://github.com/karm-patel/pyprobml-1/tree/dead_urls/notebooks/book1/01/linreg_poly_vs_degree.ipynb']
1.8 ['https://github.com/karm-patel/pyprobml-1/tree/dead_urls/notebooks/book1/01/iris_kmeans.ipynb']
1.9 ['https://github.com/karm-patel/pyprobml-1/tree/dead_urls/notebooks/book1/01/iris_pca.ipynb']
1.10 []
1.11 []
1.12 ['https://github.com/karm-patel/pyprobml-1/tree/dead_urls/notebooks/book1/01/mnist_viz_tf.ipynb', 'https://github.com/karm-patel/pyprobml-1/tree/dead_urls/notebooks/book1/01/emnist_viz_jax.ipynb']
1.13 ['https://github.com/karm-patel

In [16]:
import pandas as pd

In [25]:
df = pd.DataFrame(scripts.items(), columns=["fig", "urls"])
df.set_index("fig", drop=True, inplace=True)
df

urls
fig                                                    
1.1                                                  []
1.2                                                  []
1.3   [https://github.com/karm-patel/pyprobml-1/tree...
1.4   [https://github.com/karm-patel/pyprobml-1/tree...
1.5   [https://github.com/karm-patel/pyprobml-1/tree...
...                                                 ...
23.5                                                 []
23.6                                                 []
23.7                                                 []
23.8                                                 []
23.9                                                 []

[431 rows x 1 columns]

urls
fig                                                    
1.1                                                  []
1.2                                                  []
1.3   [https://github.com/karm-patel/pyprobml-1/tree...
1.4   [https://github.com/karm-patel/pyprobml-1/tree...
1.5   [https://github.com/karm-patel/pyprobml-1/tree...
...                                                 ...
23.5                                                 []
23.6                                                 []
23.7                                                 []
23.8                                                 []
23.9                                                 []

[431 rows x 1 columns]

In [99]:
cnt = 0
for fig in scripts:
    if scripts[fig] != []:
        cnt += len(scripts[fig])
cnt

244

In [9]:
dead_urls = []
for fig in scripts:
    for url in scripts[fig]:
        resp = requests.get(url)
        if resp.status_code != 200:
            print(fig, url)
            dead_urls.append([fig, url])

3.13 https://github.com/karm-patel/pyprobml-1/tree/dead_urls/notebooks/book1/03/mix_ber_em_mnist.ipynb
4.17 https://github.com/karm-patel/pyprobml-1/tree/dead_urls/notebooks/book1/04/gaussInferParamsMean2d.ipynb
9.5 https://github.com/karm-patel/pyprobml-1/tree/dead_urls/notebooks/book1/09/fisher_vowel_demo.ipynb


1.12: Delete emnist_viz_pytorch.ipynb from pyprobml and add emnist_viz_jax.ipynb from probml-notebooks/notebooks
3.13: Rename mix_ber_em_mnist.ipynb to mix_bernoulli_em_mnist.ipynb in this script
4.17: Rename gaussInferParamsMean2d.ipynb to gauss_infer_2d.ipynb
9.5: Rename fisher_vowel_demo.ipynb to fisher_discrim_vowel.ipynb
9.6, 9.7: Delete naive_bayes_mnist.ipynb from pyprobml and Move naive_bayes_mnist_jax.ipynb from probml-notebooks/notebooks-d2l to pyprobml
15.25: Delete positional_encoding.ipynb from pyprobml and Move positional_encoding_jax.ipynb from probml-notebooks/notebooks-d2l to pyprobml 
19.14: Move hbayes_maml.ipynb from probml-notebooks/notebooks to pyprobml
20.24, 20.25, 20.26: Move ae_mnist_conv.ipynb from probml-notebooks/notebooks to pyprobml

In [37]:
more_than_one = 0
multi_notebooks = {}
for each in scripts:

    if len(scripts[each]) > 1:
        multi_notebooks[each] = []
        print("- [ ]", each, end=": ")
        for url in scripts[each]:
            relative_path = "/".join(url.split("/")[-3:])
            print(relative_path, end=", ")
            multi_notebooks[each].append(relative_path)
        more_than_one += 1
        print()
more_than_one

- [ ] 1.12: book1/01/mnist_viz_tf.ipynb, book1/01/emnist_viz_jax.ipynb, 
- [ ] 1.13: book1/01/cifar_viz_tf.ipynb, book1/01/fashion_viz_tf.ipynb, 
- [ ] 2.2: book1/02/quantile_plot.ipynb, book1/02/gauss_plot.ipynb, 
- [ ] 2.17: book1/02/gamma_dist_plot.ipynb, book1/02/beta_dist_plot.ipynb, 
- [ ] 4.14: book1/04/dirichlet_3d_triangle_plot.ipynb, book1/04/dirichlet_3d_spiky_plot.ipynb, 
- [ ] 4.20: book1/04/logreg_iris_1d.ipynb, book1/04/logreg_iris_bayes_1d_pymc3.ipynb, 
- [ ] 5.2: book1/05/pr_plot.ipynb, book1/05/roc_plot.ipynb, 
- [ ] 5.10: book1/05/neymanPearson2.ipynb, book1/05/twoPowerCurves.ipynb, 
- [ ] 8.1: book1/08/saddle.ipynb, book1/08/extrema_fig_1d.ipynb, 
- [ ] 8.14: book1/08/newtonsMethodMinQuad.ipynb, book1/08/newtonsMethodNonConvex.ipynb, 
- [ ] 8.26: book1/08/mix_gauss_mle_vs_map.ipynb, book1/08/mix_gauss_singularity.ipynb, 
- [ ] 11.10: book1/11/ridgePathProstate.ipynb, book1/11/lassoPathProstate.ipynb, 
- [ ] 11.19: book1/11/huberLossPlot.ipynb, book1/11/linregRobustD

25

In [38]:
multi_notebooks

{'1.12': ['book1/01/mnist_viz_tf.ipynb', 'book1/01/emnist_viz_jax.ipynb'],
 '1.13': ['book1/01/cifar_viz_tf.ipynb', 'book1/01/fashion_viz_tf.ipynb'],
 '2.2': ['book1/02/quantile_plot.ipynb', 'book1/02/gauss_plot.ipynb'],
 '2.17': ['book1/02/gamma_dist_plot.ipynb', 'book1/02/beta_dist_plot.ipynb'],
 '4.14': ['book1/04/dirichlet_3d_triangle_plot.ipynb',
  'book1/04/dirichlet_3d_spiky_plot.ipynb'],
 '4.20': ['book1/04/logreg_iris_1d.ipynb',
  'book1/04/logreg_iris_bayes_1d_pymc3.ipynb'],
 '5.2': ['book1/05/pr_plot.ipynb', 'book1/05/roc_plot.ipynb'],
 '5.10': ['book1/05/neymanPearson2.ipynb', 'book1/05/twoPowerCurves.ipynb'],
 '8.1': ['book1/08/saddle.ipynb', 'book1/08/extrema_fig_1d.ipynb'],
 '8.14': ['book1/08/newtonsMethodMinQuad.ipynb',
  'book1/08/newtonsMethodNonConvex.ipynb'],
 '8.26': ['book1/08/mix_gauss_mle_vs_map.ipynb',
  'book1/08/mix_gauss_singularity.ipynb'],
 '11.10': ['book1/11/ridgePathProstate.ipynb',
  'book1/11/lassoPathProstate.ipynb'],
 '11.19': ['book1/11/huberLossP

### Create dummy notebooks for multi notebook

In [64]:
# fig_no: captions mapping
fig_numbers_stripped = list(map(lambda fig: fig.strip(), fig_numbers))
captions_dict = dict(zip(fig_numbers_stripped, captions))

In [67]:
for fig_no in multi_notebooks:
    print(fig_no, captions_dict[fig_no], end="\n\n")

1.12 (a) Visualization of the MNIST dataset. Each image is 28 × 28. There are 60k training
examples and 10k test examples. We show the first 25 images from the training set.
Generated by mnist_viz_tf.py. (b) Visualization of the EMNIST dataset. There are
697,932 training examples, and 116,323 test examples, each of size 28 × 28. There are
62 classes (a-z, A-Z, 0-9). We show the first 25 images from the training set.
Generated by emnist_viz_jax.ipynb.
19

1.13 (a) Visualization of the Fashion-MNIST dataset [XRV17]. The dataset has the same
size as MNIST, but is harder to classify. There are 10 classes: T-shirt/top, Trouser,
Pullover, Dress, Coat, Sandal, Shirt, Sneaker, Bag, Ankle-boot. We show the first 25
images from the training set. Generated by fashion_viz_tf.py. (b) Some images from
the CIFAR-10 dataset [KH09]. Each image is 32 × 32 × 3, where the final dimension of
size 3 refers to RGB. There are 50k training examples and 10k test examples. There
are 10 classes: plane, car, bird,

In [176]:
py_pattern = r"\S+?\.py"
ipynb_pattern = r"\S+?\.ipynb"
book_no = 1
for fig_no in multi_notebooks:

    caption = captions_dict[fig_no]  # caption parsed by fitz
    scripts = list(
        set(re.findall(py_pattern, caption) + re.findall(ipynb_pattern, caption))
    )  # extract script names from caption
    print(scripts, end=": ")

    # for each script, replace it with markdown-link which redirects
    # it to original notebook url
    for no, script in enumerate(scripts):
        url = os.path.join(base_url, multi_notebooks[fig_no][no])
        notebook_name = url.split("/")[-1]
        markdown_link = f"[{notebook_name}]({url})"
        caption = caption.replace(script, markdown_link)

    # remove number after caption
    last_no = caption.split()[-1]
    if last_no.isnumeric():
        caption = caption.replace(last_no, "")

    # make relative path for new dummy notebook
    chapter_no = int(fig_no.split(".")[0])
    relative_path = os.path.join(
        "../notebooks/",
    )
    dummpy_notebook = f"fig_{fig_no.replace('.','_')}.ipynb"
    fig_path = os.path.join(relative_path, f"book{book_no}/{chapter_no:02d}", dummpy_notebook)
    print(fig_path.split("/")[-1], end="\n")

    # create new notebook and add caption to it
    # https://stackoverflow.com/questions/38193878/how-to-create-modify-a-jupyter-notebook-from-code-python
    nb = nbf.v4.new_notebook()
    nb["cells"] = [nbf.v4.new_markdown_cell(caption)]
    with open(fig_path, "w") as f:
        nbf.write(nb, f)

['mnist_viz_tf.py', 'emnist_viz_jax.ipynb']: fig_1_12.ipynb
['fashion_viz_tf.py', 'cifar_viz_tf.py']: fig_1_13.ipynb
['gauss_plot.py', 'quantile_plot.py']: fig_2_2.ipynb
['beta_dist_plot.py', 'gamma_dist_plot.py']: fig_2_17.ipynb
['dirichlet_3d_spiky_plot.py', 'dirichlet_3d_triangle_plot.py']: fig_4_14.ipynb
['logreg_iris_bayes_1d_pymc3.py', 'logreg_iris_1d.py']: fig_4_20.ipynb
['pr_plot.py', 'roc_plot.py']: fig_5_2.ipynb
['twoPowerCurves.py', 'neymanPearson2.py']: fig_5_10.ipynb
['extrema_fig_1d.py', 'saddle.py']: fig_8_1.ipynb
['newtonsMethodNonConvex.py', 'newtonsMethodMinQuad.py']: fig_8_14.ipynb
['mix_gauss_mle_vs_map.py', 'mix_gauss_singularity.py']: fig_8_26.ipynb
['ridgePathProstate.py', 'lassoPathProstate.py']: fig_11_10.ipynb
['linregRobustDemoCombined.py', 'huberLossPlot.py']: fig_11_19.ipynb
['bagging_trees.py', 'rf_demo_2d.py']: fig_18_4.ipynb
['ae_mnist_conv.ipynb', 'vae_mnist_conv_lightning.ipynb']: fig_20_24.ipynb
['ae_mnist_conv.ipynb', 'vae_mnist_conv_lightning.ipynb'

In [101]:
!pip install TexSoup

  Created wheel for TexSoup: filename=TexSoup-0.3.1-py3-none-any.whl size=27666 sha256=45fa65ef18ccf8ae2ca501f628a01f2bbe20a86e98e01b27a883c46df6bcca0b
  Stored in directory: /home/patel_zeel/.cache/pip/wheels/5a/25/08/ee7ad7d9f11d425d81dd32135de9caf13e69ae8662a85fbd3c
Successfully built TexSoup
